src/v0/parity.rs

```diff
@@ -330,4 +335,97 @@ mod test {
             Err(ValidationError::Conversion)
         )
     }
+
+    #[test]
+    fn versions_encode_and_decode() {
+        let holder = [1_u8; 32];
+        let doughnut = DoughnutV0 {
+            issuer: [0_u8; 32],
+            holder,
+            domains: vec![("TestDomain".to_string(), vec![])],
+            expiry: 0,
+            not_before: 0,
+            payload_version: 0x0515,
+            signature_version: 0x1a,
+            signature: H512::default(),
+        };
+
+        let parsed_doughnut = DoughnutV0::decode(&mut &doughnut.encode()[..]).unwrap();
+        assert_eq!(parsed_doughnut.signature_version, 0x1a);
+        assert_eq!(parsed_doughnut.payload_version, 0x0515);
+    }
+
+    #[test]
+    fn payload_version_does_not_cross_contaminate() {
+        let holder = [1_u8; 32];
+        let doughnut = DoughnutV0 {
+            issuer: [0_u8; 32],
+            holder,
+            domains: vec![("TestDomain".to_string(), vec![])],
+            expiry: 0,
+            not_before: 0,
+            payload_version: 0xffff,
+            signature_version: 0x00,
+            signature: H512::default(),
+        };
+
+        let parsed_doughnut = DoughnutV0::decode(&mut &doughnut.encode()[..]).unwrap();
+        assert_eq!(parsed_doughnut.signature_version, 0x00);
+        assert_eq!(parsed_doughnut.payload_version, VERSION_11BIT_MASK);
+    }
+
+    #[test]
+    fn signature_version_does_not_cross_contaminate() {
+        let holder = [1_u8; 32];
+        let doughnut = DoughnutV0 {
+            issuer: [0_u8; 32],
+            holder,
+            domains: vec![("TestDomain".to_string(), vec![])],
+            expiry: 0,
+            not_before: 0,
+            payload_version: 0x0000,
+            signature_version: 0xff,
+            signature: H512::default(),
+        };
+
+        let parsed_doughnut = DoughnutV0::decode(&mut &doughnut.encode()[..]).unwrap();
+        assert_eq!(parsed_doughnut.payload_version, 0x0000);
+        assert_eq!(parsed_doughnut.signature_version, 0x1f);
+    }
+
+    #[test]
+    fn short_domain_name_is_parsed() {
+        let holder = [1_u8; 32];
+        let doughnut = DoughnutV0 {
+            issuer: [0_u8; 32],
+            holder,
+            domains: vec![("Smol".to_string(), vec![])],
+            expiry: 0,
+            not_before: 0,
+            payload_version: 0x0000,
+            signature_version: 0xff,
+            signature: H512::default(),
+        };
+
+        let parsed_doughnut = DoughnutV0::decode(&mut &doughnut.encode()[..]).unwrap();
+        assert_eq!(parsed_doughnut.domains[0].0, "Smol");
+    }
+
+    #[test]
+    fn long_domain_name_is_truncated() {
+        let holder = [1_u8; 32];
+        let doughnut = DoughnutV0 {
+            issuer: [0_u8; 32],
+            holder,
+            domains: vec![("SweetLikeAChic-a-CherryCola".to_string(), vec![])],
```

Comment:
just a thought we should throw an error in the `add_domain()` for users, so they don't get surprised when it gets truncated

Reply:
Yeah I agree, that would be really nice.
`encode` doesn't let us throw errors (which is a good thing), but if we add a doughnut builder, then that would be ideal! 
Doughnuts are currently just built from their structs afaict